tokio-sync/src/barrier.rs

```diff
@@ -0,0 +1,153 @@
+use crate::semaphore::{Permit, Semaphore};
+use crate::watch::{Receiver, Sender};
+use futures_util::{pin_mut, ready};
+use pin_project::pin_project;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// A barrier enables multiple threads to synchronize the beginning of some computation.
+///
+/// ```
+/// # #[tokio::main]
+/// # async fn main() {
+/// use std::sync::Arc;
+/// use tokio_sync::Barrier;
+/// use futures_util::future::join_all;
+///
+/// let mut handles = Vec::with_capacity(10);
+/// let barrier = Arc::new(Barrier::new(10));
+/// for _ in 0..10 {
+///     let c = barrier.clone();
+///     // The same messages will be printed together.
+///     // You will NOT see any interleaving.
+///     handles.push(async move {
+///         println!("before wait");
+///         let wr = c.wait().await;
+///         println!("after wait");
+///         wr
+///     });
+/// }
+/// // Will not resolve until all "before wait" messages have been printed
+/// let wrs = join_all(handles).await;
+/// // Exactly one barrier will resolve as the "leader"
+/// assert_eq!(wrs.into_iter().filter(|wr| wr.is_leader()).count(), 1);
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct Barrier {
+    semaphore: Semaphore,
+    generation: AtomicUsize,
+    waker: Sender<usize>,
+    wait: Receiver<usize>,
+}
+
+#[pin_project]
```

Comment:
I don't _think_ you need `pin_project`. I don't see how the future is not `Unpin`.